### PR TITLE
Refactor Ledger

### PR DESCRIPTION
**Pull request by vinniefalco**
Odds and ends

### CODE REVIEW

File: src/ripple/app/ledger/InboundLedger.cpp [hunk 1]
```diff
@@ -161,14 +161,15 @@ bool InboundLedger::tryLocal ()
...
+                data.data(), data.size(), true);
             getApp().getNodeStore ().store (
                 hotLEDGER, std::move (data), mHash);
         }
         else
         {
-            mLedger = std::make_shared<Ledger> (
-                strCopy (node->getData ()), true);
+            mLedger = std::make_shared<Ledger>(
+                node->getData().data(), node->getData().size(), true);
```
Comment (seelabs) [inline src/ripple/app/ledger/InboundLedger.cpp]: Would it make sense to pass around a `Slice` instead? So `getData` would return a slice and the `Ledger` constructor would take a `Slice`.

Comment (vinniefalco) [inline src/ripple/app/ledger/InboundLedger.cpp]: My intention was to avoid changing anything except the ledger-related classes. A more thorough treatment of `NodeObject` is definitely on the agenda.

Comment (seelabs) [inline src/ripple/app/ledger/InboundLedger.cpp]: What about something like `make_shared<Ledger>(toSlice(node->getData())`? Change would stay with ledger (except for the `toShared` helper classes.

Comment (vinniefalco) [inline src/ripple/app/ledger/InboundLedger.cpp]: That's even worse - adding an extra pointless function. For that, might as well just change NodeObject (which I will do, but not right now)

**Comment by HowardHinnant** [comment]
LGTM. :+1:

File: src/ripple/app/ledger/Ledger.cpp [hunk 2]
```diff
@@ -1762,7 +1694,9 @@ void Ledger::updateFees ()
     }
 
     {
-        StaticScopedLockType sl (sPendingSaveLock);
+        // VFALCO Why not do this before calling getASNode?
+        std::lock_guard<
+            std::mutex> lock(mutex_);
```
Comment (seelabs) [inline src/ripple/app/ledger/Ledger.cpp]: `std::lock_guard<decltype(mutex_)> lock(mutex_);` is an interesting idiom if you object to a type alias for the `lock_guard`. I wouldn't change this - I don't think we'll change the mutex type. We can also bug Howard about adding template type inference for constructors to the standard :-)

Comment (vinniefalco) [inline src/ripple/app/ledger/Ledger.cpp]: People shouldn't be changing `std::mutex` and more importantly we must avoid recursive mutexes.

Comment (HowardHinnant) [inline src/ripple/app/ledger/Ledger.cpp]: Or if you're really itching to get the deduction today:
````
auto lock = beast::make_lock(mutex_);
````
from https://github.com/ripple/rippled/blob/develop/src/beast/beast/utility/make_lock.h#L31

Comment (vinniefalco) [inline src/ripple/app/ledger/Ledger.cpp]: How about
```
beast::with_lock(mutex_,[&]()
{
    ...
});

File: src/ripple/app/ledger/PendingSaves.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+namespace ripple {
+
+/** Keeps track of which ledgers haven't been fully saved.
+
+    During the ledger building process this collection will keep
+    track of those ledgers that are being built but have not yet
+    been completely written.
+*/
+class PendingSaves
+{
```
Comment (seelabs) [inline src/ripple/app/ledger/PendingSaves.h]: You could make this a generic thread-safe `flat_set` templated on element type and then type alias `PendingSaves`.

File: src/ripple/app/ledger/impl/Fees.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+#include <BeastConfig.h>
+#include <ripple/app/ledger/Fees.h>
+
+namespace ripple {
+
+std::uint64_t
+scaleFeeBase (LoadFeeTrack& track,
+    Ledger const& ledger, std::uint64_t fee)
+{
+    ledger.deprecatedUpdateCachedFees();
```
Comment (seelabs) [inline src/ripple/app/ledger/impl/Fees.cpp]: Don't call `deprecatedUpdateCachedFees`. `getBaseFee` calls it. I'm going to see if we can make the `deprecatedUpdateCachedFees` private.

File: src/ripple/app/ledger/Ledger.cpp [hunk 5]
```diff
@@ -1820,7 +1739,26 @@ std::vector<uint256> Ledger::getNeededAccountStateHashes (
...
+    ledger.deprecatedUpdateCachedFees();
+    return getApp().getFeeTrack().scaleFeeBase(fee,
+        ledger.getBaseFee(), ledger.getReferenceFeeUnits());
+}
+
+std::uint64_t
+scaleFeeLoad (Ledger const& ledger,
+    std::uint64_t fee, bool admin)
+{
+    ledger.deprecatedUpdateCachedFees();
```
Comment (seelabs) [inline src/ripple/app/ledger/Ledger.cpp]: Rm `deprecatedUpdateCachedFees` call

File: src/ripple/app/ledger/Ledger.cpp [hunk 6]
```diff
@@ -1820,7 +1739,26 @@ std::vector<uint256> Ledger::getNeededAccountStateHashes (
...
-std::set<std::uint32_t> Ledger::sPendingSaves;
+//------------------------------------------------------------------------------
+
+// Converts a fee in fee units to a fee in drops
+std::uint64_t
+scaleFeeBase (Ledger const& ledger,
+    std::uint64_t fee)
+{
+    ledger.deprecatedUpdateCachedFees();
+    return getApp().getFeeTrack().scaleFeeBase(fee,
```
Comment (seelabs) [inline src/ripple/app/ledger/Ledger.cpp]: Rn `deprecatedUpdateCachedFees` call

File: src/ripple/app/ledger/Ledger.h [hunk 7]
```diff
@@ -420,6 +416,13 @@ class Ledger
         return mHash;
     }
 
+    // Updates the fees cached in the ledger.
+    // Safe to call concurrently. We shouldn't be storing
+    // fees in the Ledger object, they should be a local side-structure
+    // associated with a particular module (rpc, tx processing, consensus)
+    //
+    void deprecatedUpdateCachedFees() const;
```
Comment (seelabs) [inline src/ripple/app/ledger/Ledger.h]: Make private.

Comment (vinniefalco) [inline src/ripple/app/ledger/Ledger.h]: The whole point of this refactor was to make the function public, are you suggesting I undo the change?

Comment (seelabs) [inline src/ripple/app/ledger/Ledger.h]: I'm missing the larger picture then. It does not need to be public and a function that updates an internal cache would normally be private. Where are you going with this that it needs to be public?

File: src/ripple/app/main/Application.cpp [hunk 8]
```diff
@@ -257,6 +258,7 @@ class ApplicationImp
     NodeStoreScheduler m_nodeStoreScheduler;
     std::unique_ptr <SHAMapStore> m_shaMapStore;
     std::unique_ptr <NodeStore::Database> m_nodeStore;
+    PendingSaves pendingSaves_;
```
Comment (seelabs) [inline src/ripple/app/main/Application.cpp]: `pendingSaves_` -> `m_pendingSaves` to match the style of the rest of the file.

I think new code should conform to the coding style a file currently has and then if we change the style change it all at once.

Comment (vinniefalco) [inline src/ripple/app/main/Application.cpp]: @seelabs veto - this lets you see what was changed recently versus what is old. I rather have additional information instead of adherance to conformity for conformity's sake.

Comment (vinniefalco) [inline src/ripple/app/main/Application.cpp]: `pendingSaves_` is clearly better than `m_pendingSaves`. Are you suggesting we use an inferior style for new code just because we aren't doing a bulk search and replace?

File: src/ripple/app/ledger/Ledger.h (hunk 7 above)
Comment (vinniefalco) [inline src/ripple/app/ledger/Ledger.h]: Done

**Comment by seelabs** [comment]
:+1:

**Comment by vinniefalco** [comment]
@rec I'm going to try to remove those extra constructors entirely
